File: frontend/src/types/application.ts
export type ApplicationStatus =
  | "pending"
  | "reviewed"
  | "shortlisted"
  | "rejected"
  | "invited"
  | "accepted";
export type DocumentType =
  | "cv"
  | "diploma"
  | "cover_letter"
  | "identity_card"
  | "certificate"
  | "other";

export interface Application {
  id: number;
  job_offer_id: number;
  candidate_id: number;
  status: ApplicationStatus;
  created_at: string;
  updated_at: string;
}

export interface ApplicationWithDetails extends Application {
  job_offer: {
    id: number;
    title: string;
    company: {
      id: number;
      name: string;
    };
  };
  candidate: {
    id: number;
    full_name: string;
    email: string;
  };
  documents: Document[];
}

export interface ApplicationCreate {
  job_offer_id: number;
}

export interface ApplicationStatusUpdate {
  status: ApplicationStatus;
}

export interface Document {
  id: number;
  application_id: number;
  document_type: DocumentType;
  file_url: string;
  file_name: string;
  file_mime_type: string;
  uploaded_at: string;
}

export interface ApplicationsListResponse {
  total: number;
  applications: ApplicationWithDetails[];
}

File: backend/app/models/application.py
from datetime import datetime
from sqlalchemy import Column, Integer, String, Text, DateTime, ForeignKey
from sqlalchemy.orm import relationship

from ..database import Base


class Application(Base):
    __tablename__ = "applications"

    id = Column(Integer, primary_key=True, index=True)
    job_offer_id = Column(Integer, ForeignKey("job_offers.id"), nullable=False)
    candidate_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    cover_letter = Column(Text, nullable=True)
    status = Column(String(50), default="pending")  # pending, reviewed, shortlisted, rejected, invited, accepted
    created_at = Column(DateTime, default=datetime.utcnow)

    job_offer = relationship("JobOffer", back_populates="applications")
    candidate = relationship("User", back_populates="applications")
    documents = relationship("ApplicationDocument", back_populates="application", cascade="all, delete-orphan")

    @property
    def candidate_full_name(self):
        return self.candidate.full_name if self.candidate else None

    @property
    def candidate_email(self):
        return self.candidate.email if self.candidate else None

    @property
    def job_offer_title(self):
        return self.job_offer.title if self.job_offer else None

    @property
    def job_offer_location(self):
        return self.job_offer.location if self.job_offer else None
